### DISCUSSION TITLE
tutorials/i18n/internationalizing_games

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#252] [User-contributed notes]
# tutorials/i18n/internationalizing_games

Introduction: While indie or niche games usually do not need localization, games targeting a more massive market often require localization. Godot offers many tools to make this process more straig...

https://docs.godotengine.org/en/4.3/tutorials/i18n/internationalizing_games.html

**Comment by MajorMcDoom** [comment]
The `Translation contexts` section needs more information. It currently only mentions the value of having contexts, and demonstrates the consuming API, but it says nothing about how these contexts are actually added to the translation assets that are passed to translators.

**Reply by tetrapod00** [reply]
Hi! Could you [open an issue](https://github.com/godotengine/godot-docs/issues) on the Godot docs repo for this? We like to track issues there rather than in the comments of the docs pages.

**Reply by MajorMcDoom** [reply]
> Hi! Could you [open an issue](https://github.com/godotengine/godot-docs/issues) on the Godot docs repo for this? We like to track issues there rather than in the comments of the docs pages.

No problem, I will do that. Just for future reference: what exactly is the doc page comment section for? I was under the assumption it was to bring up things like this.

**Reply by tetrapod00** [reply]
Take a look [here](https://github.com/godotengine/godot-docs-user-notes/discussions/1). It's for sharing additional information, workarounds, or useful code snippets, or linking to useful external resources.

In practice it *is* used quite often to report issues in documentation pages, despite this being against the rules as written. This is allowed to an extent because otherwise some issues would not be reported at all. However, it's strongly preferred to report issues via the `godot-docs` repo (or to open a PR yourself, if you're able to).

**Reply by felipebottega** [reply]
A year has passed and still nothing! I was reading the tutorial and came to the conclusion that I should give up on using context because there's no indication whatsoever of how it's done. This can't stay like this...

**Reply by Calinou** [reply]
https://github.com/godotengine/godot-docs/pull/11421 documents support for contexts in CSV translations, which is a new feature in 4.6.

In 4.5 and prior, you can only supply translation contexts through [gettext PO translations](https://docs.godotengine.org/en/stable/tutorials/i18n/localization_using_gettext.html). https://github.com/godotengine/godot-docs/pull/11442 further clarifies this.

**Reply by felipebottega** [reply]
Thank you. I happened to talk to GPT and it told me that it was necessary to use PO files for this. This tutorial mentions these files in the pluralization section. I think it would also be worthwhile to mention them in the context section. At least so we have somewhere to read about it.

**Comment by perrotuerto** [comment]
I would add a link to "[Pseudolocalization](https://docs.godotengine.org/en/stable/tutorials/i18n/pseudolocalization.html)" in "Making controls resizable."

I didnt see a link to that useful section here, as I saw for other relevant sections.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11235

**Comment by MarkoGrbec** [comment]
seems a little bit undocumented and a feature is missing or I haven't found it.

make .csv file eg. translation.csv

with content eg.:
```
keys,en,sl
HELLO,Hello,Zdravo
GOODBYE,Goodbye,Nasvidenje
KAKO_SI_PROGRAMERSKO,"How are you, programmer?","Kako si, programer?"
PLAY,Play,Igra
OPTIONS,Options,Nastavitve
OI,"hello","ola"
```
this is your main translation


once it's saved right click and reimport
it'll make new files according to the languages you have in CSV

than as the doc says Project -> ProjectSettings -> Localization -> Translations (add newly made files) eg. translation.en.translation

than you are ready to translate:

```
	TranslationServer.set_locale("sl")
	print(tr("OI")) # ola
	
	TranslationServer.set_locale("en")
	print(tr("OI")) # hello
```
missing feature or I haven't found it
and make next file a singleton: ReverseTranslation.gd
```
extends Node
var translation_map = {}

func _ready():
	var file = FileAccess.open("res://translation.csv", FileAccess.READ)
	if file:
		var languages = file.get_csv_line().slice(1)  # ["en", "sl"]
		
		for lang in languages:
			translation_map[lang] = {}
		
		while not file.eof_reached():
			var line = file.get_csv_line()
			if line.size() > 0 and line[0] != "":
				var key = line[0]
				for i in range(languages.size()):
					var lang = languages[i]
					if i+1 < line.size() and line[i+1] != "":
						var translation = line[i+1]
						translation_map[lang][translation] = key
		file.close()

func get_key_by_translation(text, locale) -> String:
	return translation_map.get(locale, {}).get(text, "")
```

hope this helps anyone in their journey

**Reply by Calinou** [reply]
> and a feature is missing or I haven't found it.

What feature are you looking for? Check methods available in the [TranslationServer](https://docs.godotengine.org/en/stable/classes/class_translationserver.html) singleton; it's possible that what you're looking for is already implemented there.

**Reply by MarkoGrbec** [reply]
reverse to find a key from translation.

the one I have implemented up there in comment.

**Comment by Pinaz993** [comment]
I feel like I'm missing something obvious here, so if I am, please point me to the relevant section.

How does one add a translation key and its associated value in a given locale? I'm an `en_us` guy, but I'd like my game to be translatable from the get-go, so I'm looking to add in translation keys for any strings that are user-facing. I'm seeing how to import a CSV to automatically add key-value pairs, but not how to tell the engine that I want to add a specific key.

I also looked in Project Settings > Localization, and the only way I see to add a translation is to import a `.CSV`. Do I need to make the `.CSV` first, or is there a way to manually add a key and values?

**Reply by Calinou** [reply]
CSVs cannot automatically be generated [like POT files can be](https://docs.godotengine.org/en/stable/tutorials/i18n/localization_using_gettext.html#automatic-generation-using-the-editor). You need to manually create a CSV file, then enter translation keys in the CSV to match the strings used in your project.

**Reply by Pinaz993** [reply]
So you're saying there's no GUI way to add translation keys? If I were more comfortable with C, I'd fork and open a pull request, but I've only used GC languages. I'll have to settle for a feature request. Either way, I'll make a CSV when I have some energy to use.
